cmd/notify-postgresql.go

```diff
@@ -0,0 +1,208 @@
+			"Table was not specified in Notify config for" +
+				" PostgreSQL")
+	}
+
+	connStr := pgN.ConnectionString
+	// check if connection string is specified
+	if connStr == "" {
+		// build from other parameters
+		params := []string{}
+		if pgN.Host != "" {
+			params = append(params, "host="+pgN.Host)
+		}
+		if pgN.Port != "" {
+			params = append(params, "port="+pgN.Port)
+		}
+		if pgN.User != "" {
+			params = append(params, "user="+pgN.User)
+		}
+		if pgN.Password != "" {
+			params = append(params, "password="+pgN.Password)
+		}
+		if pgN.Database != "" {
+			params = append(params, "dbname="+pgN.Database)
+		}
+		connStr = strings.Join(params, " ")
+	}
+
+	db, err := sql.Open("postgres", connStr)
+	if err != nil {
+		return pgConn{}, err
+	}
+
+	// ping to check that server is actually reachable.
+	err = db.Ping()
+	if err != nil {
+		return pgConn{}, err
+	}
+
+	// TODO: check that table exists?
+
+	// create prepared statements
+	stmts := make(map[string]*sql.Stmt)
+	// insert or update statement
+	stmts["upsertRow"], err = db.Prepare(fmt.Sprintf(upsertRow, pgN.Table))
+	if err != nil {
+		return pgConn{}, errors.New("Unable to create upsert prepared statement for postgres notification.")
+	}
+	stmts["deleteRow"], err = db.Prepare(fmt.Sprintf(deleteRow, pgN.Table))
+	if err != nil {
+		return pgConn{}, errors.New("Unable to create delete prepared statement for postgres notification.")
+	}
+
+	return pgConn{connStr, pgN.Table, stmts, db}, nil
+}
+
+func newPostgreSQLNotify(accountID string) (*logrus.Logger, error) {
+	pgNotify := serverConfig.GetPostgreSQLNotifyByID(accountID)
+
+	// Dial postgres
+	pgC, err := dialPostgreSQL(pgNotify)
+	if err != nil {
+		return nil, err
+	}
+
+	pgLog := logrus.New()
+
+	pgLog.Out = ioutil.Discard
+
+	pgLog.Formatter = new(logrus.JSONFormatter)
+
+	pgLog.Hooks.Add(pgC)
+
+	return pgLog, nil
+}
+
+func (pgC pgConn) Close() {
+	// we first close all prepared statements
+	for _, v := range pgC.preparedStmts {
+		v.Close()
+	}
+	// close db connection
+	pgC.DB.Close()
+}
+
+func (pgC pgConn) Fire(entry *logrus.Entry) error {
+	// get event type by trying to convert to string
+	entryEventType, ok := entry.Data["EventType"].(string)
+	if !ok {
+		// TODO: why is this error silently ignored (in other
+		// notification agents too)?
+		return nil
+	}
+
+	// Check if the event delete
+	if eventMatch(entryEventType, []string{"s3:ObjectRemoved:*"}) {
+		_, err := pgC.preparedStmts["deleteRow"].Exec(entry.Data["Key"])
+		if err != nil {
+			return fmt.Errorf(
+				"Error deleting event with key = %v - got postgres error - %v",
+				entry.Data["Key"], err,
+			)
+		}
+	} else {
+		// json encode the value for the row
+		value, err := json.Marshal(map[string]interface{}{
+			"Records": entry.Data["Records"],
+		})
+		if err != nil {
+			return fmt.Errorf(
+				"Unable to encode event %v to JSON - got error - %v",
+				entry.Data["Records"], err,
+			)
+		}
+
+		// attempt to insert into the table
```

Comment:
Fixed this comment. It is atomic, and failures are only related to connection problems and other un-controllables.